File: backend/internal/worker/agent/providers/kimi/todo.go
package kimi

import (
	"encoding/json"

	"github.com/leapmux/leapmux/generated/contracts"

	"github.com/leapmux/leapmux/internal/worker/todoevents"
)

// kimiTodoStatus maps a TodoList status word onto the neutral status. Kimi
// spells a finished item `done`, which the shared parser does not know.
func kimiTodoStatus(word string) todoevents.Status {
	switch word {
	case contracts.KimiTodoStatusPending:
		return todoevents.StatusPending
	case contracts.KimiTodoStatusInProgress:
		return todoevents.StatusInProgress
	case contracts.KimiTodoStatusDone:
		return todoevents.StatusCompleted
	default:
		return todoevents.StatusFromProviderWord(word)
	}
}

// ExtractTodoEvent reads Kimi Code's to-do list off a TodoList call.
//
// Each call states the WHOLE list in `args.todos`, so this is a snapshot. A call
// that omits `todos` only reads the list and changes nothing, and an empty list
// clears it (docs/en/reference/tools.md). The list rides the call's opening
// `tool.call.started`; its result repeats nothing the list needs.
func (kimiProvider) ExtractTodoEvent(spanType string, content []byte, _ func() []byte) (todoevents.Event, bool) {
	if spanType != contracts.KimiToolTodoList {
		return todoevents.Event{}, false
	}
	var call struct {
		Type string `json:"type"`
		Name string `json:"name"`
		Args struct {
			Todos *[]struct {
				Title  string `json:"title"`
				Status string `json:"status"`
			} `json:"todos"`
		} `json:"args"`
	}
	if err := json.Unmarshal(content, &call); err != nil ||
		call.Type != contracts.KimiEventToolCallStarted || call.Name != contracts.KimiToolTodoList || call.Args.Todos == nil {
		return todoevents.Event{}, false
	}
	items := make([]todoevents.Item, 0, len(*call.Args.Todos))
	for _, todo := range *call.Args.Todos {
		items = append(items, todoevents.Item{Content: todo.Title, Status: kimiTodoStatus(todo.Status)})
	}
	return todoevents.Event{Kind: todoevents.KindSnapshot, Snapshot: items}, true
}

File: frontend/src/components/chat/providers/kimi/extractors/todo.ts
import type { TodoItem } from '~/models/todo'
import { KIMI_TODO_STATUS } from '~/generated/contracts/kimi-protocol'
import { isObject, pickString } from '~/lib/jsonPick'
import { todoRowKey } from '~/models/todo'
import { normalizeTodoStatus } from '../../../normalizers/todo'

/**
 * One `TodoList` status word, in the shared status vocabulary.
 *
 * Kimi Code spells a finished item `done`, which the shared normalizer does not know.
 */
function kimiTodoStatus(status: unknown): TodoItem['status'] {
  switch (status) {
    case KIMI_TODO_STATUS.Done:
      return 'completed'
    case KIMI_TODO_STATUS.InProgress:
      return 'in_progress'
    case KIMI_TODO_STATUS.Pending:
      return 'pending'
    default:
      return normalizeTodoStatus(status)
  }
}

/**
 * The items of a `TodoList` call's `todos` argument.
 *
 * Each call states the WHOLE list, so position is the identity of an item: Kimi Code
 * gives an item no id of its own. An item is `{title, status}`.
 */
export function kimiTodoItems(raw: unknown): TodoItem[] {
  if (!Array.isArray(raw))
    return []
  return raw.flatMap((entry, index) => {
    if (!isObject(entry))
      return []
    const content = pickString(entry, 'title')
    return [{ rowKey: todoRowKey(undefined, index, content), content, status: kimiTodoStatus(entry.status), activeForm: '' }]
  })
}
